libraries/AP_Landing/AP_Landing_Deepstall.cpp

```diff
@@ -0,0 +1,380 @@
+        SRV_Channels::set_output_scaled(SRV_Channel::k_rudder, output*4500);
+
+    } else {
+        // allow the normal servo control of the channel
+        SRV_Channels::set_output_scaled(SRV_Channel::k_aileron_with_input,
+                                        SRV_Channels::get_output_scaled(SRV_Channel::k_aileron));
+    }
+
+    // hand off rudder control to deepstall controlled
+    return true;
+}
+
+bool AP_Landing::type_deepstall_request_go_around(void)
+{
+    flags.commanded_go_around = true;
+    return true;
+}
+
+bool AP_Landing::type_deepstall_is_throttle_suppressed(void) const
+{
+    return type_deepstall_stage == DEEPSTALL_STAGE_LAND;
+}
+
+bool AP_Landing::type_deepstall_get_target_altitude_location(Location &location)
+{
+    memcpy(&location, &type_deepstall_landing_point, sizeof(Location));
+    return true;
+}
+
+int32_t AP_Landing::type_deepstall_get_target_airspeed_cm(void) const
+{
+    if (type_deepstall_stage == DEEPSTALL_STAGE_APPROACH ||
+        type_deepstall_stage == DEEPSTALL_STAGE_LAND) {
+        return pre_flare_airspeed * 100;
+    } else {
+        return aparm.airspeed_cruise_cm;
+    }
+}
+
+const DataFlash_Class::PID_Info& AP_Landing::type_deepstall_get_pid_info(void) const
+{
+    return type_deepstall_PID.get_pid_info();
+}
+
+void AP_Landing::type_deepstall_build_approach_path(void)
+{
+    Vector3f wind = ahrs.wind_estimate();
+    // TODO: Support a user defined approach heading
+    type_deepstall_target_heading_deg = (degrees(atan2f(-wind.y, -wind.x)));
+
+    memcpy(&type_deepstall_extended_approach, &type_deepstall_landing_point, sizeof(Location));
+    memcpy(&type_deepstall_arc_exit, &type_deepstall_landing_point, sizeof(Location));
+
+    //extend the approach point to 1km away so that there is always a navigational target
+    location_update(type_deepstall_extended_approach, type_deepstall_target_heading_deg, 1000.0);
+
+    float expected_travel_distance = type_deepstall_predict_travel_distance(wind, type_deepstall_landing_point.alt / 100);
+    float approach_extension = expected_travel_distance + type_deepstall_approach_extension;
+    // an approach extension of 0 can result in a divide by 0
+    if (fabsf(approach_extension) < 1.0f) {
+        approach_extension = 1.0f;
+    }
+
+    location_update(type_deepstall_arc_exit, type_deepstall_target_heading_deg + 180, approach_extension);
+    memcpy(&type_deepstall_arc, &type_deepstall_arc_exit, sizeof(Location));
+    memcpy(&type_deepstall_arc_entry, &type_deepstall_arc_exit, sizeof(Location));
+
+    // TODO: Support loitering on either side of the approach path
+    location_update(type_deepstall_arc, type_deepstall_target_heading_deg + 90.0, aparm.loiter_radius);
+    location_update(type_deepstall_arc_entry, type_deepstall_target_heading_deg + 90.0, aparm.loiter_radius * 2);
+
+    GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "Arc: %3.8f %3.8f",
```

Comment:
would be nice for these strings to start with "DS" or "Landing" so we know where they come from